legion/engine/llri-vk/source.cpp

```diff
@@ -1,14 +1,265 @@
-#include <llri/llri.h>
+#include <llri/llri.hpp>
+
+#include <vector>
+#include <map>
+#include <vulkan/vulkan.hpp>
 
 namespace legion::graphics::llri
 {
-    namespace detail
+    namespace internal
+    {
+        std::map<std::string, vk::LayerProperties>& queryAvailableLayers();
+        std::map<std::string, vk::ExtensionProperties>& queryAvailableExtensions();
+        Result mapVkResult(const vk::Result& result);
+    }
+
+    Result createInstance(const InstanceDesc& desc, Instance* instance)
+    {
+        if (instance == nullptr)
+            return Result::ErrorInvalidUsage;
+        if (desc.numExtensions > 0 && desc.extensions == nullptr)
+            return Result::ErrorInvalidUsage;
+
+        auto* result = new InstanceT();
```

Comment:
if `instance` was a reference you also wouldn't need to do any dynamic allocation. but if you rly want it to be a pointer, then i would still recommend either using a `shared_ptr` or a `unique_ptr`. ever since C++11 there is no use to actually using the `new` keyword anymore. It is actually discouraged to ever use the `new` keyword since in almost all cases code designed not to use it will be less bug/leak prone and/or faster.

Reply:
I don't think a low level API should be pushing any form of RAII by default. If the user wants to wrap this in a smart pointer then they're free to

Reply:
fair, but it could still be a reference instead of a pointer. you are now restricting the use of this function to being dynamically managed using new instead of maybe some external stack managed memory or other ways of managing memory.

Reply:
Libraries often solve this by allowing the user to pass custom malloc/free functions along, I'd be willing to implement something like that, but I think that should be dedicated to a separate PR